tests/freshness.rs

```diff
@@ -172,16 +172,192 @@ fn changing_features_is_ok() {
+            authors = []
+
+            [features]
+            ftest  = []
+        "#)
+        .file("dep_crate/src/lib.rs", r#"
+            #[cfg(feature = "ftest")]
+            pub fn yo() {
+                println!("ftest on")
+            }
+            #[cfg(not(feature = "ftest"))]
+            pub fn yo() {
+                println!("ftest off")
+            }
+        "#)
+        .file("a/Cargo.toml", r#"
+            [package]
+            name    = "a"
+            version = "0.0.1"
+            authors = []
+
+            [dependencies]
+            dep_crate = {path = "../dep_crate", features = []}
+        "#)
+        .file("a/src/lib.rs", "")
+        .file("a/src/main.rs", r#"
+            extern crate dep_crate;
+            use dep_crate::yo;
+            fn main() {
+                yo();
+            }
+        "#)
+        .file("b/Cargo.toml", r#"
+            [package]
+            name    = "b"
+            version = "0.0.1"
+            authors = []
+
+            [dependencies]
+            dep_crate = {path = "../dep_crate", features = ["ftest"]}
+        "#)
+        .file("b/src/lib.rs", "")
+        .file("b/src/main.rs", r#"
+            extern crate dep_crate;
+            use dep_crate::yo;
+            fn main() {
+                yo();
+            }
+        "#);
+
+    /* Build a/ package once */
+    assert_that(p.cargo_process("run").cwd(p.root().join("a")),
+                execs().with_status(0)
+                       .with_stdout("ftest off")
+                       .with_stderr("\
+[..]Compiling dep_crate v0.0.1 ([..])
+[..]Compiling a v0.0.1 ([..])
+[FINISHED] debug [unoptimized + debuginfo] target(s) in [..]
+[RUNNING] `[..]target/debug/a`
+"));
+
+    /* Rebuild a/ package. Shouldn't cause recompile */
+    assert_that(p.cargo("run").cwd(p.root().join("a")),
+                execs().with_status(0)
+                       .with_stdout("ftest off")
+                       .with_stderr("\
+[FINISHED] debug [unoptimized + debuginfo] target(s) in [..]
+[RUNNING] `[..]target/debug/a`
+"));
+
+    /* Build b/ package. Should recompile deps_crate because of different feature flags */
+    assert_that(p.cargo("run").cwd(p.root().join("b")),
+                execs().with_status(0)
+                       .with_stdout("ftest on")
+                       .with_stderr("\
+[..]Compiling dep_crate v0.0.1 ([..])
+[..]Compiling b v0.0.1 ([..])
+[FINISHED] debug [unoptimized + debuginfo] target(s) in [..]
+[RUNNING] `[..]target/debug/b`
+"));
+
+    /* Rebuild b/ package. Should not recompile since we just built b */
+    assert_that(p.cargo("run").cwd(p.root().join("b")),
+                execs().with_status(0)
+                       .with_stdout("ftest on")
+                       .with_stderr("\
+[FINISHED] debug [unoptimized + debuginfo] target(s) in [..]
+[RUNNING] `[..]target/debug/b`
+"));
+
+    /* Build a/ package again. If we cache different feature dep builds correctly,
```

Comment:
Hm so this test is passing?

If so, I think that may be due to how Cargo tracks where if one target is rebuilt it will unconditionally rebuild all transitive dependants. You can perhaps try to expose this through:

```
cd a
cargo build -p dep_crate
cargo run
```

That'll force Cargo to try to understand that the dependency was rebuilt, maybe? (not sure if that'd fail)